lib/tbot/ssh_proxy.go

```diff
@@ -0,0 +1,262 @@
+package tbot
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"io"
+	"log/slog"
+	"net"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/gravitational/trace"
+	"golang.org/x/crypto/ssh/agent"
+	"google.golang.org/grpc"
+
+	"github.com/gravitational/teleport/api/client"
+	"github.com/gravitational/teleport/api/client/proto"
+	proxyclient "github.com/gravitational/teleport/api/client/proxy"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/utils/grpc/interceptors"
+	libclient "github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/teleport/lib/client/identityfile"
+	"github.com/gravitational/teleport/lib/resumption"
+	"github.com/gravitational/teleport/lib/tbot/config"
+	"github.com/gravitational/teleport/lib/tbot/identity"
+	"github.com/gravitational/teleport/lib/tbot/tshwrap"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// ProxySSHConfig contains configuration parameters required
+// to initialize the local ssh proxy.
+type ProxySSHConfig struct {
+	BotConfig                 *config.BotConfig
+	TSHConfigPath             string
+	ProxyServer               string
+	Cluster                   string
+	UserHostPort              string
+	EnableResumption          bool
+	TLSRoutingEnabled         bool
+	ConnectionUpgradeRequired bool
+	Log                       *slog.Logger
+}
+
+// ProxySSH creates a local ssh proxy, dialing a node and transferring data through
+// stdin and stdout, to be used as an OpenSSH/PuTTY proxy command.
+func ProxySSH(ctx context.Context, proxyConfig ProxySSHConfig) error {
+	tshConfig, err := libclient.LoadTSHConfig(proxyConfig.TSHConfigPath)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	user, hostPort, ok := strings.Cut(proxyConfig.UserHostPort, "@")
+	if !ok {
+		hostPort = proxyConfig.UserHostPort
+	}
+
+	proxy := proxyConfig.ProxyServer
+	cluster := proxyConfig.Cluster
+	expanded, matched := tshConfig.ProxyTemplates.Apply(hostPort)
+	if matched {
+		proxyConfig.Log.DebugContext(ctx, "proxy templated matched", "populated_template", expanded)
+		if expanded.Cluster != "" {
+			cluster = expanded.Cluster
+		}
+		if expanded.Proxy != "" {
+			proxy = expanded.Proxy
+		}
+		if expanded.Host != "" {
+			hostPort = expanded.Host
+		}
+	}
+
+	proxyHost, _, err := net.SplitHostPort(proxy)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	facade, keyring, err := parseIdentity(proxyConfig.BotConfig, proxyHost, cluster)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	sshConfig, err := facade.SSHClientConfig()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if user != "" {
+		sshConfig.User = user
+	}
+
+	if cluster == "" {
+		cluster = facade.Get().ClusterName
+	}
+
+	pclt, err := proxyclient.NewClient(ctx, proxyclient.ClientConfig{
+		ProxyAddress:      proxyConfig.ProxyServer,
```

Comment:
Should the proxy template be allowed to override this?
```suggestion
		ProxyAddress:      proxy,
```